### ISSUE TITLE
defaultProps and object destructuring defaults are treated differently

### ISSUE THREAD

**Issue by FezVrasta** [facebook/flow#5752]
https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6gEoCmAhgMZ6RayqgDkWJ5tA3IoqdMQM5egCCoQgA8chAHYATXkTI4AdAGFqmMeJwAeAN6hIsWAC5QXHFmRiA5qAC+APlCbEASGPEcyUqAmFIxAK7QcAAUqdF4AXntEUGjQR10DOgAjYixaABoo2KtWJwZJQiwACgBKSOjHRwYcXywxUBwAC2QuOXQQlvjWRytEHsRIXzFyZFg6gCFC7XjQCNpk1OtDKb1DY1MLa1KHcqqaus7etgGht1HQABVCYxKy2N3a0ELM8vUJZAA3G0yKx3UBYC+MSBsXUY1AAPBYAKVCw31+wDenyijmKrD6-UGwzOl2MACYbts7oRqg8nkDfojAbEKn9wVTgSCwfEwgAieYsumQ0BiWCCLAwuHqBEfQEotFAA

Basically, `defaultProps` is properly interpreted (or not?) by Flow, and then if you define a property with a default value, you don't have to mark it as optional, because a value will always be present.

Object destructuring, on the other hand, will ask you to mark the property as optional even if you assign a default value to it

**Comment by apsavin** [comment]
Just want to note that you can write your code like [this](https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6gEoCmAhgMZ6RayqgDkWJ5tA3IoqdMQM5egCCoQgA8chAHYATXkTI4AdAGFqmMeJwAeAN6hIsWAC5QXHFmRiA5qAC+APlCbEASGPEcyUqAmFIxAK7QcAAUqdF4AXntEUGjQR10DOgAjYixaABoo2KtWJwZJQiwACgBKSOjHRwYcXywxUBwAC2QuOXQQlvjWRytEHsRIXzFyZFg6gCFC7XjrQym9Q2NTC2tSh3KqmrrO3rYxuS8ffyD20Ai1nXmklPTenIGht1HQABVCYxKy2I3a0ELM8vUEmQADcbJkKo51AJgGCYnDYuoxqAYciwGJYIIsFQsODIcAgaCoo5iqw+v1BsMnq9jAAmD7nSqEao-P5wyEE2GxCpQ5Gc+EIpHxMIAImSWGFvNRoHRmOxuPU+JBsOJpKAA)

**Comment by FezVrasta** [comment]
Yes that's an alternative, but since I already use object destructuring it makes much sense to use the default values feature of it.

**Comment by apsavin** [comment]
I think there's a difference: 

for `React.Component` when you type `Props` you describe a type of a `this.props`, not of an interface. So flow calculates the type of an interface of the component using this information and `defaultProps`.

for a function, when you type a function parameter, you type function's interface in first place.

From my point of view, it would be better to change the way how we work with components and to start type their interface instead of `props` - but I'm not sure it can be reached in practice.

**Comment by daneden** [comment]
Came here to +1! It doesn’t seem quite right that I can’t use Object destructuring default values as an alternative to `defaultProps`. I expect the following to run without errors:

```js
type Props = {
  value: boolean
}

function Thing({value = true}: Props) {
  return <div>The value is {value ? 'true' : 'false'}!</div>
}

function App() {
  return (
    <div>
      <Thing value={false} />
      <Thing />
    </div>
  )
}
```

This currently fails with Flow 0.96.1: [Test case](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuYASgKYCGAxkVHnJmAOR4XWOqoECe2pYACnWwBnMAF4wAb1RgwAN3IwArqQBcYAEZw4MCgDtUAXw5Qle6ujh6wAFQAW6PQHMAFJIXK+EgnhWH1gnAiAJRSMmAsBEp41gA8ACbocgB89nweKmDoou6KmQD8TD4qjGDqjFCKwqSMhgCEscCJKUYmZhZWYACC2NguodKykdHWLuGyCUnJ47JgsfaOTvJ5pGKSlTDVhmDA07Oz8w7OO3sHTVPhwUZAA).

And passes when using `defaultProps`: [Test case](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuYASgKYCGAxkVHnJmAOR4XWOqoECe2pYACnWwBnMAF4wAb1RgwAN3IwArqQBcYAEZw4MCgDtUAXw5Qle6ujh6wAFQAW6PQHMAFJIXLSh9YLgiAlFIyYCwESnjWADwAJuhyAHz2fB4qYOii7oqpAPxMBHgqjGDqjFCKwqSMhgCEkcCxCUYc9o5OAHTRpGVKMAS+IuJBsilqYPkqTaim5gSW1gCC2NgugdKyoeHWLsGyMXHxO7JgkS3O8lmkYpJlMBWGYMAHR0cnDmePh3UNT2D+RkA).

**Comment by pascalduez** [comment]
@daneden My understanding is `defaultProps` and default parameters are handled differently by Flow, which makes sense after all. While `defaultProps` should be marked as required since Flow will do it's `$Diff` thing, default parameter should be marked as optional. Flow will understand there's a default and don't require you to define check those params.

**Comment by SamChou19815** [comment]
Expected behavior
